### DISCUSSION TITLE
Deployments clean up when a Pull Request is merged

### DISCUSSION THREAD

**Discussion by jgoux** [vercel/community#761] [Help] [answered]
Hello all,

I'm working on a little integration between Vercel and [Snaplet](https://www.snaplet.dev/) leveraging Github Actions and the Vercel Git Integration.

I noticed that when a pull request is merged and the branch is deleted, the deployments aren't deleted, either on Vercel or Github's side. Is there any way to automatically clean up all the deployments related to a branch on both sides?

**Comment by amyegan** [comment]
You can [remove deployments using the CLI](https://vercel.com/docs/cli#commands/remove) or  [Vercel API](https://vercel.com/docs/rest-api#endpoints/deployments/delete-a-deployment). The downside is you need to keep track of which deployments are related to that branch if you don't want to remove preview deployments for other branches. If you want to delete all deployments aside from active preview or production, then you can use the [safe option](https://vercel.com/docs/cli#commands/remove/unique-options/safe) like `vercel remove <my-project-name> --safe`.

Vercel doesn't provide a way to remove branches or change data in GitHub, but this sounds like something you could add to your GitHub Actions.

**Reply by jgoux** [reply]
Thanks @amyegan, I'll do something custom then. 👍

**Reply by amyegan** [reply]
Wanted to add a link to our guide on [how to use GitHub Actions with Vercel](https://vercel.com/support/articles/how-can-i-use-github-actions-with-vercel) in case that helps with your custom solution

**Reply by jgoux** [reply]
Thanks for the doc, it's very helpful!

Is there any way of deploying to Vercel manually (using the vercel cli as in the doc you shared) and still benefitting from the Vercel GitHub App for the bot comments + the GitHub deployment status? It would mean disabling the Vercel GitHub App ability to deploy when opening a PR and when adding new commits to a PR. But in a perfect world, I'd still want it to deploy for the production branch. 😄

I came across: https://vercel.com/docs/project-configuration#git-configuration, it would be the perfect place to have a way of telling the Github App to not trigger deployments for a given environment (in my case the "preview" environment), but still sharing the deployment status + commenting in the PR! 🤞

I'm not sure of the behavior of the https://vercel.com/docs/project-configuration#git-configuration/github-enabled option, is it disabling all the functionalities of the Github App or just the deployment? If it's just the deployment it would be perfect in my case, but I'd like to go further and have it scoped to an environment:
```json
{
  "github": {
    "enabled:" ["production"]
  }
}
```

**Reply by amyegan** [reply]
You can use "Ignored Build Step" with the Vercel GitHub integration to control which branches are automatically built and which to skip https://vercel.com/support/articles/how-do-i-use-the-ignored-build-step-field-on-vercel

**Reply by jgoux** [reply]
Yes, this is actually what I implemented in my solution, I wanted to be sure I wasn't missing something more simple.
Thanks again for your support!

**Reply by jgoux** [reply]
One last thing sorry 😅 
Even with the "Ignored Build Step", if I deploy with the vercel cli manually using `vercel deploy --prebuilt --prod --token=${{ secrets.VERCEL_TOKEN }}`, is there a way to distinguish between a deployment triggered by the CLI versus the GitHub App? In my use case, I'd reject only the deployment targeting the `preview` environment **and** being initiated by the GitHub App. I need to accept the deployments targeting the `preview` environment and triggered by the vercel cli.

**Reply by jgoux** [reply]
Solved, I can inject my own metadata using `--meta` 👍

**Reply by jgoux** [reply]
So...it's not **really** solved yet 😄
The workflow is working for the deployment part, but I have no feedback from the GitHub App.
It seems like using `vercel deploy --prebuilt --token=$VERCEL_ACCESS_TOKEN` isn't triggering any event related to the GitHub App so I can't benefit from the GitHub App feedbacks about the deployments. 😢
Is there any option to `vercel deploy --prebuilt` to act as if it was a deploy hook related to a branch? Or do I have to go back to my more complicated solution involving creating deploy hooks dynamically and discard `vercel deploy --prebuilt`?

**Reply by jgoux** [reply]
I ended up using the deploy hook approach to be fully integrated with the GitHub App. 👌
FYI I've made a reusable action to take control of Vercel's deployment when dealing with the GitHub App: https://github.com/snaplet/vercel-action

**Reply by amyegan** [reply]
This is awesome! Thanks for sharing 😁 🎉

**Comment by leerob** [comment, marked as answer]
You can now configure the retention period for deployments through the dashboard and CLI.

For example, canceled and errored deployments might be set to 30 days retention, while production deployments might be set to 1 year. Recently deleted deployments are shown in your project settings and can be instantly restored within 30 days of deletion.

Learn more in our [documentation](https://vercel.com/docs/security/deployment-retention).
